File: Protocol/Responses/LoginResponse.cpp
#include "LoginResponse.h"

LoginResponse::LoginResponse(LoginResult result, string authToken)
{
    this->authToken = authToken;
    this->result = result;
}

string LoginResponse::GetAuthToken() const
{
    return this->authToken;
}

LoginResult LoginResponse::GetResult() const
{
    return this->result;
}

ResponseType LoginResponse::GetResponseType() const
{
    return ResponseType::LoginResult;
}

string LoginResponse::ToString() const
{
    ostringstream stream;
    stream << (int)GetResponseType() << ' ' << (int)GetResult() << ' ' << GetAuthToken();
    return stream.str();
}

LoginResponse *LoginResponse::Parse(string str)
{
    istringstream stream(str);
    int responseType, result;
    string authToken;
    stream >> responseType >> result >> authToken;

    if ((ResponseType)responseType != ResponseType::LoginResult)
        throw "Invalid type of request";

    return new LoginResponse((LoginResult)result, authToken);
}

File: Client/Protocol/Responses/LoginResponse.cs
using System;
using System.IO;
using System.Text;

namespace NetTalk.Protocol.Responses
{
    public enum LoginResult
    {
        Ok,
        InvalidCredentials,
    };

    public class LoginResponse : ResponseBase
    {
        public LoginResponse(LoginResult result, string authToken)
        {
            Result = result;
            AuthToken = authToken;
        }

        public string AuthToken { get; private set; }
        public override ResponseType ResponseType { get; } = ResponseType.LoginResult;
        public LoginResult Result { get; private set; }

        public static LoginResponse Parse(byte[] bytes)
        {
            using (MemoryStream stream = new MemoryStream(bytes))
            using (BinaryReader reader = new BinaryReader(stream))
            {
                ResponseType responseType = (ResponseType)reader.ReadInt32();
                if (responseType != ResponseType.LoginResult)
                    throw new InvalidOperationException("Invalid type of response");

                LoginResult result = (LoginResult)reader.ReadInt32();
                string authToken = Encoding.ASCII.GetString(reader.ReadBytes(reader.ReadInt32()));

                return new LoginResponse(result, authToken);
            }
        }
    }
}